examples/04_Variational_and_Approximate_GPs/Natural_Gradient_Descent.ipynb

```diff
@@ -0,0 +1,376 @@
+       "model_id": "cf4264ce654e49b0956fafad188850bc",
+       "version_major": 2,
+       "version_minor": 0
+      },
+      "text/plain": [
+       "HBox(children=(FloatProgress(value=0.0, description='Epoch', max=4.0, style=ProgressStyle(description_width='i…"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "data": {
+      "application/vnd.jupyter.widget-view+json": {
+       "model_id": "",
+       "version_major": 2,
+       "version_minor": 0
+      },
+      "text/plain": [
+       "HBox(children=(FloatProgress(value=0.0, description='Minibatch', max=13.0, style=ProgressStyle(description_wid…"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "data": {
+      "application/vnd.jupyter.widget-view+json": {
+       "model_id": "",
+       "version_major": 2,
+       "version_minor": 0
+      },
+      "text/plain": [
+       "HBox(children=(FloatProgress(value=0.0, description='Minibatch', max=13.0, style=ProgressStyle(description_wid…"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "data": {
+      "application/vnd.jupyter.widget-view+json": {
+       "model_id": "",
+       "version_major": 2,
+       "version_minor": 0
+      },
+      "text/plain": [
+       "HBox(children=(FloatProgress(value=0.0, description='Minibatch', max=13.0, style=ProgressStyle(description_wid…"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "data": {
+      "application/vnd.jupyter.widget-view+json": {
+       "model_id": "",
+       "version_major": 2,
+       "version_minor": 0
+      },
+      "text/plain": [
+       "HBox(children=(FloatProgress(value=0.0, description='Minibatch', max=13.0, style=ProgressStyle(description_wid…"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "model.train()\n",
+    "likelihood.train()\n",
+    "mll = gpytorch.mlls.VariationalELBO(likelihood, model, num_data=train_y.size(0))\n",
+    "\n",
+    "num_epochs = 1 if smoke_test else 4\n",
+    "epochs_iter = tqdm.notebook.tqdm(range(num_epochs), desc=\"Epoch\")\n",
+    "for i in epochs_iter:\n",
+    "    minibatch_iter = tqdm.notebook.tqdm(train_loader, desc=\"Minibatch\", leave=False)\n",
+    "    \n",
+    "    for x_batch, y_batch in minibatch_iter:\n",
+    "        ### Perform NGD step to optimize variational parameters\n",
+    "        variational_ngd_optimizer.zero_grad()\n",
+    "        output = model(x_batch)\n",
+    "        loss = -mll(output, y_batch)\n",
+    "        minibatch_iter.set_postfix(loss=loss.item())\n",
+    "\n",
+    "        # Multiply the loss by the amount of data in order to take the appropriate step size!\n",
+    "        (loss * mll.num_data).backward()\n",
```

Comment:
What do you think about making like a 5 line `class NGD(torch.optim.SGD):` that takes `num_data` and wraps this behavior (e.g., does the multiply on the gradients in step before calling super)? The other primary advantage would be avoiding the trap of using other optimizers, since then we're being pretty explicit about what to use.

Reply:
Yeah. That's actually a really good idea.

Reply:
Okay - pushed.